sources/slack/__init__.py

```diff
@@ -0,0 +1,114 @@
+from typing import Any, Dict, Iterable, Optional
+
+import dlt
+from dlt.common.time import ensure_pendulum_datetime
+from dlt.common.typing import TAnyDateTime, TDataItem
+from dlt.extract.source import DltResource
+from pendulum import DateTime
+
+from .helpers import resource_creator
+from .settings import DEFAULT_START_DATE, MAX_PAGE_SIZE
+
+
+@dlt.source(name="slack")
+def slack_source(
+    access_token: str = dlt.secrets.value,
+    page_size: int = MAX_PAGE_SIZE,
+    start_date: Optional[TAnyDateTime] = DEFAULT_START_DATE,
+    end_date: Optional[TAnyDateTime] = None,
+) -> Iterable[DltResource]:
+    """
+    The source for the Slack pipeline. Available resources are conversations and access_logs.
+
+    Args:
+        access_token: the oauth access_token used to authenticate.
+        page_size: The max number of items to fetch per page. Defaults to 1000.
+        start_date: The start time of the range for which to load. Defaults to January 1st 2000.
+        end_date: The end time of the range for which to load data.
+
+    Returns:
+        Iterable[DltResource]: A list of DltResource objects representing the data resources.
+    """
+
+    def convert_dt(dt: TAnyDateTime, to_ts: bool = False) -> Any:
+        if dt is None:
+            return None
+        out_dt = ensure_pendulum_datetime(dt)
+        if to_ts:
+            return out_dt.timestamp()
+        return out_dt
+
+    end_dt: Optional[DateTime] = convert_dt(end_date)
+    start_dt: Optional[DateTime] = convert_dt(start_date)
+
+    def get_convesation_child(
+        item: Dict[str, Any],
+        resource: str,
+        response_path: str,
+        parent_path: Dict[str, str],
+        end_date: Optional[TAnyDateTime] = None,
+        start_date: dlt.sources.incremental[TAnyDateTime] = dlt.sources.incremental(
+            "ts",
+            initial_value=start_dt,
+            allow_external_schedulers=True,
+        ),
+    ) -> Iterable[TDataItem]:
+        start_date_ts = convert_dt(start_date.last_value, to_ts=True)
+        end_date_ts = convert_dt(end_date, to_ts=True)
+        params = {
+            "oldest": start_date_ts,
+            "latest": end_date_ts,
+        }
+        extra_params = {k: item[v] for k, v in parent_path.items()}
+        params.update(extra_params)
+        yield resource_creator(
+            access_token=access_token,
+            resource=resource,
+            params=params,
+            response_path=response_path,
+            parent_context=params,
+            datetime_fields=["ts"],
+        )
+
+    conversations = dlt.resource(
+        resource_creator,
+        name="conversations",
+        primary_key="id",
+        write_disposition="replace",
+    )(
+        access_token=access_token,
+        resource="conversations.list",
+        response_path="channels",
+        datetime_fields=["updated", "created"],
+        page_size=page_size,
+    )
+
+    yield conversations
+
+    yield conversations | dlt.transformer(
```

Comment:
OK this is really smart and could be used to create a DAG where we learn about channels at execution time and create tables for each channels. in case of slack this is more complicated though. I think we are better off if we follow the mongodb pattern:
- get a list of channels by evaluating list of conversations
- for each conversation (or if an explicit list of channel names/ids is present - filter only those on the list) create a dlt resource which will contain conversation history. give it meaningful name ie. the one you use below (but I think it is too long. maybe channel_{channel} is engough and please use channel name not id)
- and now the trick: channel history contains many different event types. route each event type to a separate tables! I sent you links to examples in previous review

so at the end:
- you have a list of resources for each conversation
- each resource routes data to many tables corresponding to event types

Reply:
The list of conversations is the list of channels, using the admin token, it could bring other conversations using a user token, from what I could find doing some tests.

For event types in the `conversation.history`, they are all messages, there is no separation of event, not sure how to break it. And this `client_msg_id` is not always present, endpoints that get reaction, for example get it based on ts and channel.

Sample history:

```
        {
            "client_msg_id": "afa46452-4716-4c40-938a-a1cf398ee503",
            "type": "message",
            "text": "hey, I've got a question regarding the new mongodb source, does it also work for documentdb?",
            "user": "U05A0SCG6KT",
            "ts": "1693837655.609649",
            "blocks": [
                {
                    "type": "rich_text",
                    "block_id": "ZaxQX",
                    "elements": [
                        {
                            "type": "rich_text_section",
                            "elements": [
                                {
                                    "type": "text",
                                    "text": "hey, I've got a question regarding the new mongodb source, does it also work for documentdb?"
                                }
                            ]
                        }
                    ]
                }
            ],
            "team": "T04DHMAF13Q",
            "thread_ts": "1693837655.609649",
            "reply_count": 5,
            "reply_users_count": 4,
            "latest_reply": "1693844421.520359",
            "reply_users": [
                "U054H6GT3BP",
                "U05LJG8UV2R",
                "U05A0SCG6KT",
                "U04DQ9VUN1Y"
            ],
            "is_locked": false,
            "subscribed": false
        },
```